app/prism/Prism.scala

```diff
@@ -37,11 +38,23 @@ object Prism {
     } yield instances
   }
 
-  def instancesWithAmis(stackStageApp: SSA)(implicit config: AMIableConfig, ec: ExecutionContext): Attempt[List[(Instance, Option[AMI])]] = {
+  private def instancesAndAmis(stackStageApp: SSA)(implicit config: AMIableConfig, ec: ExecutionContext): Attempt[(List[Instance], List[AMI])] = {
     for {
       prodInstances <- getInstances(stackStageApp)
       amiAttempts = amiArns(prodInstances).map(getAMI)
       amis <- Attempt.successfulAttempts(amiAttempts)
-    } yield instanceAmis(prodInstances, amis)
+    } yield (prodInstances, amis)
+  }
+
+  def instancesWithAmis(stackStageApp: SSA)(implicit config: AMIableConfig, ec: ExecutionContext): Attempt[List[(Instance, Option[AMI])]] = {
```

Comment:
It's just a question of taste so it's up to you, but this file is supposed to just contain the impure code that fetches data from Prism. The logic for handling the pure value that gets returned from Prism lives elsewhere. This helps to enforce the discipline of separating easily testable logic from code that depends on side effects and external systems.

`instancesWithAmis` and the other functions here now can't be tested without mocking the Prism API and providing an execution context. If they took a Prism response as an argument instead, you'd be able to test them by passing them a pure value and testing the pure value it returns. Of course, in these examples the code is so trivial that it's very likely you'll say they don't need tests. While this is fine and it's clear that there's no great harm caused by introducing these functions here, it comes back to the question of discipline. This starts the slippery slide of mixing logic with impure calls which will encourage future developers to add more features here, manipulating Prism responses in the same blocks that fetch those responses.

I'm happy to chat about this IRL, and definitely convince-able both ways, but thought it worth sharing the ideas behind the design as it stands.